pq-crypto/sike_r3/sikep434r3_kem.c

```diff
@@ -0,0 +1,142 @@
+    ephemeralsk[SIKE_P434_R3_SECRETKEY_A_BYTES - 1] &= SIKE_P434_R3_MASK_ALICE;
+
+    // Encrypt
+    EphemeralKeyGeneration_A(ephemeralsk, ct);
+    EphemeralSecretAgreement_A(ephemeralsk, pk, jinvariant);
+    shake256(h, SIKE_P434_R3_MSG_BYTES, jinvariant, SIKE_P434_R3_FP2_ENCODED_BYTES);
+    for (int i = 0; i < SIKE_P434_R3_MSG_BYTES; i++) {
+        ct[i + SIKE_P434_R3_PUBLIC_KEY_BYTES] = temp[i] ^ h[i];
+    }
+
+    // Generate shared secret ss <- H(m||ct)
+    memcpy(&temp[SIKE_P434_R3_MSG_BYTES], ct, SIKE_P434_R3_CIPHERTEXT_BYTES);
+    shake256(ss, SIKE_P434_R3_SHARED_SECRET_BYTES, temp, SIKE_P434_R3_CIPHERTEXT_BYTES+SIKE_P434_R3_MSG_BYTES);
+
+    return 1;
+}
+
+// SIKE's decapsulation
+// Input:   secret key sk         (SIKE_P434_R3_SECRET_KEY_BYTES = SIKE_P434_R3_MSG_BYTES + SIKE_P434_R3_SECRETKEY_B_BYTES + SIKE_P434_R3_PUBLIC_KEY_BYTES bytes)
+//          ciphertext message ct (SIKE_P434_R3_CIPHERTEXT_BYTES = SIKE_P434_R3_PUBLIC_KEY_BYTES + SIKE_P434_R3_MSG_BYTES bytes)
+// Outputs: shared secret ss      (SIKE_P434_R3_SHARED_SECRET_BYTES bytes)
+int sike_p434_r3_crypto_kem_dec(unsigned char *ss, const unsigned char *ct, const unsigned char *sk)
+{
+    unsigned char ephemeralsk_[SIKE_P434_R3_SECRETKEY_A_BYTES];
+    unsigned char jinvariant_[SIKE_P434_R3_FP2_ENCODED_BYTES];
+    unsigned char h_[SIKE_P434_R3_MSG_BYTES];
+    unsigned char c0_[SIKE_P434_R3_PUBLIC_KEY_BYTES];
+    unsigned char temp[SIKE_P434_R3_CIPHERTEXT_BYTES+SIKE_P434_R3_MSG_BYTES];
+
+    // Decrypt
+    EphemeralSecretAgreement_B(sk + SIKE_P434_R3_MSG_BYTES, ct, jinvariant_);
+    shake256(h_, SIKE_P434_R3_MSG_BYTES, jinvariant_, SIKE_P434_R3_FP2_ENCODED_BYTES);
+    for (int i = 0; i < SIKE_P434_R3_MSG_BYTES; i++) {
+        temp[i] = ct[i + SIKE_P434_R3_PUBLIC_KEY_BYTES] ^ h_[i];
+    }
+
+    // Generate ephemeralsk_ <- G(m||pk) mod oA
+    memcpy(&temp[SIKE_P434_R3_MSG_BYTES], &sk[SIKE_P434_R3_MSG_BYTES + SIKE_P434_R3_SECRETKEY_B_BYTES], SIKE_P434_R3_PUBLIC_KEY_BYTES);
+    shake256(ephemeralsk_, SIKE_P434_R3_SECRETKEY_A_BYTES, temp, SIKE_P434_R3_PUBLIC_KEY_BYTES+SIKE_P434_R3_MSG_BYTES);
+    ephemeralsk_[SIKE_P434_R3_SECRETKEY_A_BYTES - 1] &= SIKE_P434_R3_MASK_ALICE;
+
+    // Generate shared secret ss <- H(m||ct), or output ss <- H(s||ct) in case of ct verification failure
+    EphemeralKeyGeneration_A(ephemeralsk_, c0_);
+
+    // Verify ciphertext.
+    // If c0_ and ct are NOT equal, decaps failed and we overwrite the shared secret
+    // with pseudorandom noise (ss = H(s||ct)) by performing the copy (dont_copy = false).
+    //
+    // If c0_ and ct are equal, then decaps succeeded and we skip the overwrite and output
+    // the actual shared secret: ss = H(m||ct) (dont_copy = true).
+    int dont_copy = !(CRYPTO_memcmp(c0_, ct, SIKE_P434_R3_PUBLIC_KEY_BYTES));
+    constant_time_copy_or_dont(temp, sk, SIKE_P434_R3_MSG_BYTES, dont_copy);
```

Comment:
That's an artifact of how it [appears in s2n](https://github.com/aws/s2n-tls/blob/cce65a92f6491dc7c2167c1621cf65d074970c96/pq-crypto/sike_r3/sikep434r3_kem.c#L106-L107). Since s2n already had constant time compare/copy functions, we preferred to use those instead of bringing in near-identical duplicates from the official SIKE repo. Jyoti used the SIKE code in s2n as the basis for this PR, and was able to identify `CRYPTO_memcmp` within awslc as a replacement for the constant time compare, but we couldn't find any equivalent copy function. We could bring over `ct_cmov` instead of `constant_time_copy_or_dont` - pretty sure they're nearly identical.

Reply:
ok, I see. I'd prefer if we took SIKE's implementation just because it's third party code and every change we make we kind of take on the responsibility to maintain it.

Reply:
done!